### PR TITLE
Update login/sign-up API

### LINKED ISSUE #3: Refugee Registration
As a Refugee, I can register myself on the platform with an email, password, and name.

Please create an API route `/registration` that accepts a name, password, and email.
Validate the submitted fields, email address should be unique verify that no user exists with the email already.
The name and password should not be empty.
When the fields are validated save them to the user model, make sure the model supports saving these fields.

Return a 200 response when a new User was created and representative other responses for the appropriate errors. 
When the registration was successful also log in the user.

### LINKED ISSUE #5: Refugee authentication
As a Refugee, I can register myself on the platform with an email, password, and name.

Please create an API route /authenticate that accepts a password, and email.
Validate the submitted fields, the email and password should match a saved record.

If no match was found return a 404 response indicating we could not find the user. 

Return a 200 response when we found the user and the credentials match. Create a token for the user and return this token in the response.

### PR DESCRIPTION
**Pull request by HazemBittar**
Please check login/sign-up and logout RESTful API using jwt-authentication
closes #3 closes #5

### CODE REVIEW

File: api/models/User.js [hunk 1]
```diff
@@ -1,21 +1,97 @@
...
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+const confiq = require('../config/config').get(process.env.NODE_ENV);
+const salt = 10;
+
+const userSchema = mongoose.Schema({
+    name: {
+        type: String,
+        required: true,
+        unique: true,
```
Comment (abelRoland) [inline api/models/User.js]: I am not sure if the name field should be unique.... Two "John Doe" is not a problem, the field that is unique generally is email.

File: api/models/User.js [hunk 2]
```diff
@@ -1,21 +1,97 @@
...
+userSchema.methods.deleteToken = function(token, cb) {
+    var user = this;
+
+    user.update({ $unset: { token: 1 } }, function(err, user) {
+        if (err) return cb(err);
+        cb(null, user);
+    })
+}
+
+
```
Comment (abelRoland) [inline api/models/User.js]: I am not sure if all of this belongs to here... Normally the model file is just a model for the dB, all the rest related with the model is done in external files.

**Review by abelRoland — CHANGES_REQUESTED** [review]

**Review by Sayed94h — COMMENTED** [review]
We should think about it:
What should be really unique?
As we see in another social media app: 
Your have `name`, `family name` and **username** which is unique and you can use as your login credential as well. If there are five Bobs and we want to display their names beside their stories. It is not important if they don't have unique name but if the listeners click on their name and visit their profile then they should see a unique **username** maybe combination of their `firstname`  and `lastname` or if it is still not unique we should add some number or his/her birthdate on it. Ex: Bob Becker --> bobbecker1980.

I hope it is clear. 
Nice work @HazemBittar When other reviewers approve the PR I will approve it as well. I asked other backend developers to review your PR.
Have great times

File: api/models/User.js (hunk 1 above)
Comment (Toinne) [reply api/models/User.js]: I agree shouldn't be unique

Comment (HazemBittar) [reply api/models/User.js]: Solved!

File: api/models/User.js (hunk 2 above)
Comment (Toinne) [reply api/models/User.js]: I would suggest the same, in a big application.
You don't want to mix auth logic and token within the model those are 2 different concepts.

However, this is a small enough application and I don't see the problem here. 
Mongoose provides this easy-to-use way for this sorta use case.

File: api/db/mongoose.js [hunk 3]
```diff
@@ -1,17 +1,16 @@
 require('dotenv').config();
 const mongoose = require('mongoose');
-const seedUsers = require('../scripts/seed')
 
 const uri = process.env.DB_URI
 mongoose.connect(uri, {useNewUrlParser: true, useUnifiedTopology: true});
-seedUsers();
```
Comment (Toinne) [inline api/db/mongoose.js:7]: FIX: we don't want to seed users on every connect, each time we restart the server we make DB calls not needed. This is why we use the sync.js script.

File: api/routes/login.js [hunk 4]
```diff
@@ -1,69 +1,105 @@
...
+        newUser.save((err, doc) => {
+            if (err) {
+                return res.status(400).json({
+                    error: true,
+                    message: err.message
+                });
+            }
+            res.status(200).json({
+                success: true,
+                user: { // Control the output we don't want to return the password that would be bad
```
Comment (Toinne) [inline api/routes/login.js:35]: FIX: this was an important change NEVER expose the password to the frontend. Even though it is hashed this poses a security risk. We make it way harder for the hacker to get the hashes if he needs access to the server db.

File: api/routes/login.js [hunk 5]
```diff
@@ -1,69 +1,105 @@
...
+app.post('/login', function(req, res) {
+    let token = req.cookies.auth;
+    User.findByToken(token, (err, user) => {
+        if (err) {
+            return res(err);
+        }
+
+        if (user) {
+            return res.status(400).json({
+                error: true,
```
Comment (Toinne) [inline api/routes/login.js:56]: FIX: cleaned error messages a bit. Keep them consistent each error should have a message keep the same structure throughout the application.

File: api/routes/login.js [hunk 6]
```diff
@@ -1,69 +1,105 @@
...
-app.get('/home', function (req, res) {
-  res.send('Welcome!');
-});
 
-app.get('/secret', withAuth, function (req, res) {
-  res.send('You are visiting a protected page.');
+// login user
+app.post('/login', function(req, res) {
+    let token = req.cookies.auth;
+    User.findByToken(token, (err, user) => {
```
Comment (Toinne) [inline api/routes/login.js:49]: Would say this is a bit unconventional normally you just verify the token and that is enough.
It expires after a set amount of time and you are the only party that can sign and verify it with your secret.

No need to save it on the model.

If you do want to save it for security reason, giving you control to destroy sessions I would suggest a different collection.

But no need to change now this works for now.

**Review by Toinne — APPROVED** [review]

**Review by Sayed94h — APPROVED** [review]

**Comment by Eltayeb-Elgaali** [comment]
Many changes have been done by @Toinne .
So i merge this PR.
If someone see that we need to do new changes we can do it through new ticket.

**Pull request merged**